File: testbed_ruby/test/associations/optional_one_to_m_star_test.rb
require 'test/unit'
require 'src/student_f'
require 'src/mentor_f'
require 'src/program_f'

module CruiseTest
class OptionalOneToMStarTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentF.new
    assert_equal(nil,s.get_mentor)
  end
  
  def test_ConstructorWithTooFew
  	assert_raise RuntimeError do #"Unable to create MentorF, must have at least 2 students"
      s = StudentF.new
      MentorF.new([s])
    end
  end

  def test_constructorJustBigEnough
    s = StudentF.new
    s2 = StudentF.new

    m = MentorF.new([s,s2])

    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
  end
  
  def test_constructorWatchOutForDuplicateEntries
  	assert_raise RuntimeError do #"Unable to create MentorF, must have at least 2 students"
      s = StudentF.new
      MentorF.new([s,s])
    end
  end
  
  def test_constructorCheckForExistingMentorNotEnoughToSurvive
  	assert_raise RuntimeError do #"Unable to create MentorF, must have at least 2 students"
      s1 = StudentF.new
      s2 = StudentF.new
      s3 = StudentF.new
      s4 = StudentF.new
    
      m = MentorF.new([s1,s2,s3])

      assert_equal(3, m.number_of_students)
      assert_equal(m, s1.get_mentor)
      assert_equal(m, s2.get_mentor)
      assert_equal(m, s3.get_mentor)    
      assert_equal(nil, s4.get_mentor)    

      MentorF.new([s1,s2,s4])
    end
  end   
  
  def test_constructorCheckForExistingMentorFnoughToShare
    s1 = StudentF.new
    s2 = StudentF.new
    s3 = StudentF.new
    s4 = StudentF.new
    s5 = StudentF.new
    
    m = MentorF.new([s1,s2,s3])

    assert_equal(3, m.number_of_students)
    assert_equal(m, s1.get_mentor)
    assert_equal(m, s2.get_mentor)
    assert_equal(m, s3.get_mentor)    
    assert_equal(nil, s4.get_mentor)    
    assert_equal(nil, s5.get_mentor)

    m2 = MentorF.new([s2,s4,s5])

    assert_equal(2, m.number_of_students)
    assert_equal(3, m2.number_of_students)
    assert_equal(m, s1.get_mentor)
    assert_equal(m2, s2.get_mentor)
    assert_equal(m, s3.get_mentor)    
    assert_equal(m2, s4.get_mentor)    
    assert_equal(m2, s5.get_mentor)
  end   

  def test_set_students_lessThanM
    s = StudentF.new
    s2 = StudentF.new    
    s3 = StudentF.new
    s4 = StudentF.new
    s5 = StudentF.new
    s6 = StudentF.new

    m = MentorF.new([s,s2])
    
    assert_equal(false, m.set_students([s]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(nil,s3.get_mentor)
    assert_equal(nil,s4.get_mentor)    
    assert_equal(nil,s5.get_mentor)    
    assert_equal(nil,s6.get_mentor)    
  end

  def test_set_students_aLot
    allStudents = []
    
    for i in 1..100
      allStudents << StudentF.new
    end
    
    m = MentorF.new(allStudents)
    assert_equal(100,m.number_of_students)
  end
  
  

  def test_set_students_atLeastM
  
    s = StudentF.new
    s2 = StudentF.new    
    s3 = StudentF.new
    s4 = StudentF.new

    m = MentorF.new([s,s2])
    
    assert_equal(true, m.set_students([s2, s3, s4]))
    assert_equal(3,m.number_of_students)
    assert_equal(nil,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
    assert_equal(m,s4.get_mentor)    
  end
  
  def test_add_student
  
    s = StudentF.new
    s2 = StudentF.new    
    s3 = StudentF.new
    s4 = StudentF.new
    s5 = StudentF.new
    s6 = StudentF.new

    m = MentorF.new([s,s2,s3,s4])

    assert_equal(true,m.add_student(s5))
    
    assert_equal(5,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
    assert_equal(m,s4.get_mentor)    
    assert_equal(m,s5.get_mentor)    

    assert_equal(true,m.add_student(s6))
    assert_equal(6,m.number_of_students)
  
  end  

  def test_add_studentHasExistingMentor
  
    s = StudentF.new
    s2 = StudentF.new    
    s3 = StudentF.new
    s4 = StudentF.new
    s5 = StudentF.new
    s6 = StudentF.new

    m = MentorF.new([s,s2,s3])
    m2 = MentorF.new([s4,s5,s6])
    
    assert_equal(true,m.add_student(s4))
    
    assert_equal(4,m.number_of_students)
    assert_equal(2,m2.number_of_students)
    
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
    assert_equal(m,s4.get_mentor)    
    assert_equal(m2,s5.get_mentor)    
    assert_equal(m2,s6.get_mentor)

    assert_equal(false,m.add_student(s5))
  end  

  def test_remove_student
  
    s = StudentF.new
    s2 = StudentF.new    
    s3 = StudentF.new
    s4 = StudentF.new

    m = MentorF.new([s,s2,s3])
    
    assert_equal(false,m.remove_student(s4))
    assert_equal(true,m.remove_student(s3))
    assert_equal(2,m.number_of_students)
    
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(nil,s3.get_mentor)
    assert_equal(nil,s4.get_mentor)    

    assert_equal(false,m.remove_student(s2))
    assert_equal(2,m.number_of_students)
  end  

  def test_set_students_empty
    s = StudentF.new
    s2 = StudentF.new
    s3 = StudentF.new
    m = MentorF.new([s,s2,s3])

    assert_equal(false, m.set_students([]))
    assert_equal(3,m.number_of_students)
  end
  
  def test_deletStudentDeletesTheMentorIfNotEnoughLeft
    s1 = StudentF.new
    s2 = StudentF.new

    m = MentorF.new([s1,s2])
    studentP = ProgramF.new
    mentorP = ProgramF.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)

    s1.delete
    
    assert_equal(0,m.number_of_students)
    assert_equal(nil,m.get_program)
    assert_equal(nil,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(nil,s1.get_mentor)
  end  
  
  def test_deleteStudentLeavesMentorAloneIfHasEnough
    s1 = StudentF.new
    s2 = StudentF.new
    s3 = StudentF.new

    m = MentorF.new([s1,s2,s3])
    studentP = ProgramF.new
    mentorP = ProgramF.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)

    s1.delete
    
    assert_equal(2,m.number_of_students)
    assert_equal(mentorP,m.get_program)
    assert_equal(m,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(nil,s1.get_mentor)
  end 
  
end
end

File: testbed_php/test/associations/OptionalOneToMStarTest.php
<?php

class OptionalOneToMStarTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentF();
    $this->assertEqual(null,$s->getMentor());
  }
  

  public function test_ConstructorWithTooFew()
  {
  	$this->expectException(new Exception("Unable to create MentorF, must have at least 2 students"));
    $s = new StudentF();
    new MentorF(array($s));
  }
  
  

  public function test_constructorJustBigEnough()
  {
    $s = new StudentF();
    $s2 = new StudentF();

    $m = new MentorF(array($s,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
  }
  

  public function test_constructorWatchOutForDuplicateEntries()
  {
  	$this->expectException(new Exception("Unable to create MentorF, must have at least 2 students"));
    $s = new StudentF();
    new MentorF(array($s, $s));
  }
  
  public function test_constructorCheckForExistingMentorNotEnoughToSurvive()
  {
  	$this->expectException(new Exception("Unable to create MentorF, must have at least 2 students"));
    $s1 = new StudentF();
    $s2 = new StudentF();
    $s3 = new StudentF();
    $s4 = new StudentF();
    
    $m = new MentorF(array($s1,$s2,$s3));

    $this->assertEqual(3, $m->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor());
    $this->assertEqual($m, $s2->getMentor());
    $this->assertEqual($m, $s3->getMentor());    
    $this->assertEqual(null, $s4->getMentor());    

    new MentorF(array($s1,$s2,$s4));
  }   
  

  public function test_constructorCheckForExistingMentorFnoughToShare()
  {
    $s1 = new StudentF();
    $s2 = new StudentF();
    $s3 = new StudentF();
    $s4 = new StudentF();
    $s5 = new StudentF();
    
    $m = new MentorF(array($s1,$s2,$s3));

    $this->assertEqual(3, $m->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor());
    $this->assertEqual($m, $s2->getMentor());
    $this->assertEqual($m, $s3->getMentor());    
    $this->assertEqual(null, $s4->getMentor());    
    $this->assertEqual(null, $s5->getMentor());

    $m2 = new MentorF(array($s2,$s4,$s5));

    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual(3, $m2->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor());
    $this->assertEqual($m2, $s2->getMentor());
    $this->assertEqual($m, $s3->getMentor());    
    $this->assertEqual($m2, $s4->getMentor());    
    $this->assertEqual($m2, $s5->getMentor());
  }   
  
  

  public function test_setStudents_lessThanM()
  {
    $s = new StudentF();
    $s2 = new StudentF();    
    $s3 = new StudentF();
    $s4 = new StudentF();
    $s5 = new StudentF();
    $s6 = new StudentF();

    $m = new MentorF(array($s, $s2));
    
    $this->assertEqual(false, $m->setStudents(array($s)));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual(null,$s3->getMentor());
    $this->assertEqual(null,$s4->getMentor());    
    $this->assertEqual(null,$s5->getMentor());    
    $this->assertEqual(null,$s6->getMentor());    

  }


  public function test_setStudents_aLot()
  {
    $allStudents = array();
    
    for ($i=0; $i<100; $i++)
    {
      $allStudents[] = new StudentF();
    }
    
    $m = new MentorF($allStudents);
    $this->assertEqual(100,$m->numberOfStudents());
  }
  
  

  public function test_setStudents_atLeastM()
  {
    $s = new StudentF();
    $s2 = new StudentF();    
    $s3 = new StudentF();
    $s4 = new StudentF();

    $m = new MentorF(array($s, $s2));
    
    $this->assertEqual(true, $m->setStudents(array($s2, $s3, $s4)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
    $this->assertEqual($m,$s4->getMentor());    
  }
  

  public function test_addStudent()
  {
    $s = new StudentF();
    $s2 = new StudentF();    
    $s3 = new StudentF();
    $s4 = new StudentF();
    $s5 = new StudentF();
    $s6 = new StudentF();

    $m = new MentorF(array($s,$s2,$s3,$s4));

    $this->assertEqual(true,$m->addStudent($s5));
    
    $this->assertEqual(5,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
    $this->assertEqual($m,$s4->getMentor());    
    $this->assertEqual($m,$s5->getMentor());    

    $this->assertEqual(true,$m->addStudent($s6));
    $this->assertEqual(6,$m->numberOfStudents());
  
  }  


  public function test_addStudentHasExistingMentor()
  {
    $s = new StudentF();
    $s2 = new StudentF();    
    $s3 = new StudentF();
    $s4 = new StudentF();
    $s5 = new StudentF();
    $s6 = new StudentF();

    $m = new MentorF(array($s,$s2,$s3));
    $m2 = new MentorF(array($s4,$s5,$s6));
    
    $this->assertEqual(true,$m->addStudent($s4));
    
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
    $this->assertEqual($m,$s4->getMentor());    
    $this->assertEqual($m2,$s5->getMentor());    
    $this->assertEqual($m2,$s6->getMentor());

    $this->assertEqual(false,$m->addStudent($s5));
  }  


  public function test_removeStudent()
  {
    $s = new StudentF();
    $s2 = new StudentF();    
    $s3 = new StudentF();
    $s4 = new StudentF();

    $m = new MentorF(array($s,$s2,$s3));
    
    $this->assertEqual(false,$m->removeStudent($s4));
    $this->assertEqual(true,$m->removeStudent($s3));
    $this->assertEqual(2,$m->numberOfStudents());
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual(null,$s3->getMentor());
    $this->assertEqual(null,$s4->getMentor());    

    $this->assertEqual(false,$m->removeStudent($s2));
    $this->assertEqual(2,$m->numberOfStudents());
    
    
  }  
  
  
  
  

  public function test_setStudents_empty()
  {
    $s = new StudentF();
    $s2 = new StudentF();
    $s3 = new StudentF();
    $m = new MentorF(array($s,$s2,$s3));

    $this->assertEqual(false, $m->setStudents(array()));
    $this->assertEqual(3,$m->numberOfStudents());
  }
  

  public function test_deletStudentDeletesTheMentorIfNotEnoughLeft()
  {
    $s1 = new StudentF();
    $s2 = new StudentF();

    $m = new MentorF(array($s1,$s2));
    $studentP = new ProgramF();
    $mentorP = new ProgramF();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(null,$m->getProgram());
    $this->assertEqual(null,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(null,$s1->getMentor());
  }  
  

  public function test_deleteStudentLeavesMentorAloneIfHasEnough()
  {
    $s1 = new StudentF();
    $s2 = new StudentF();
    $s3 = new StudentF();

    $m = new MentorF(array($s1,$s2,$s3));
    $studentP = new ProgramF();
    $mentorP = new ProgramF();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($mentorP,$m->getProgram());
    $this->assertEqual($m,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(null,$s1->getMentor());
  } 

  
  
}
